File: frontend/types/generation.ts
export type GenerationStatus =
  | "PENDING"
  | "QUEUED"
  | "GENERATING"
  | "RETRYING"
  | "DONE"
  | "FAILED"
  | "CANCELLED";

export interface Generation {
  request_id: string;
  conversation_id: string;
  status: GenerationStatus;
  queue_position?: number | null;
  requested_model?: string | null;
  actual_model?: string | null;
  attempt_count: number;
  error_code?: string | null;
  error_message?: string | null;
  user_message_id?: string | null;
  assistant_message_id?: string | null;
}

export interface ChatJobAccepted {
  request_id: string;
  status: GenerationStatus;
  queue_position: number | null;
}

export interface GenerationEvent {
  id: string;
  event: "queued" | "generating" | "delta" | "retrying" | "completed" | "failed";
  data: Record<string, unknown>;
}

File: backend/app/schemas/generation.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator


GenerationStatus = Literal[
    "PENDING", "QUEUED", "GENERATING", "RETRYING", "DONE", "FAILED", "CANCELLED"
]


class ChatJobRequest(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)

    conversation_id: UUID
    message: str = Field(min_length=1)
    client_request_id: UUID
    model: Literal["default", "advanced"] = "default"

    @field_validator("message")
    @classmethod
    def reject_local_command(cls, value: str) -> str:
        if value.startswith("/"):
            raise ValueError("Lệnh chat phải được xử lý cục bộ, không gửi tới model")
        return value


class GenerationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    request_id: UUID = Field(validation_alias="id")
    conversation_id: UUID
    status: GenerationStatus
    queue_position: int | None = None
    requested_model: str | None = None
    actual_model: str | None = None
    attempt_count: int
    error_code: str | None = None
    error_message: str | None = None
    user_message_id: UUID | None = None
    assistant_message_id: UUID | None = None
    created_at: datetime
    updated_at: datetime


class ChatJobAccepted(BaseModel):
    request_id: UUID
    status: GenerationStatus
    queue_position: int | None
